parser/src/lexer.rs

```diff
@@ -1397,6 +1461,183 @@ mod tests {
         }
     }
 
+    macro_rules! test_jupyter_magic_line_continuation_eol {
+        ($($name:ident: $eol:expr,)*) => {
+            $(
+            #[test]
+            fn $name() {
+                let source = format!("%matplotlib \\{}  --inline", $eol);
+                let tokens = lex_jupyter_source(&source);
+                assert_eq!(
+                    tokens,
+                    vec![
+                        Tok::MagicCommand { value: "matplotlib   --inline".to_string(), kind: MagicKind::Magic },
+                    ]
+                )
+            }
+            )*
+        };
+    }
+
+    test_jupyter_magic_line_continuation_eol! {
+        test_jupyter_magic_line_continuation_windows_eol: WINDOWS_EOL,
+        test_jupyter_magic_line_continuation_mac_eol: MAC_EOL,
+        test_jupyter_magic_line_continuation_unix_eol: UNIX_EOL,
+    }
+
+    #[test]
+    fn test_empty_jupyter_magic() {
+        let source = "%\n%%\n!\n!!\n?\n??\n/\n,\n;";
+        let tokens = lex_jupyter_source(source);
+        assert_eq!(
+            tokens,
+            vec![
+                Tok::MagicCommand {
+                    value: "".to_string(),
+                    kind: MagicKind::Magic,
+                },
+                #[cfg(feature = "full-lexer")]
+                Tok::NonLogicalNewline,
```

Comment:
What's the reasoning that this is a non logical newline. Isn't it a logical newline, because it *terminates* a statement? 

Can we add a test where a magic command uses an invalid indent.

Reply:
> What's the reasoning that this is a non logical newline. Isn't it a logical newline, because it terminates a statement?

My main reason being that as these tokens are filtered out before parsing, they should end with a `NonLogicalNewline` as otherwise there'll be multiple newline tokens i.e., multiple blank lines. This is similar to the `Comment` token. 

Although, now that I think of it, without the `full-lexer` feature (but with `Mode::Jupyter`), the `NonLogicalNewline` won't be filtered out while the `MagicCommand` token will be.

> Can we add a test where a magic command uses an invalid indent.

Do you mean something like the following?
```python
for i in range(10):
    print('hello')
   !pwd
```

Reply:
> Do you mean something like the following?

Yes, exactly. Sorry, I should have provided an example.

Reply:
We've decided to move ahead with the current implementation but will have to consider indentation once the parser is updated to account for any indentation error.